src/components/App.js

```diff
@@ -10,16 +10,85 @@ import MainPage from "./MainPage";
 import { getNews } from "../utils/newsApi";
 import MobileMenu from "./MobileMenu";
 import ActivePopupContext from "../contexts/ActivePopupContext";
+import CurrentUserContext from "../contexts/CurrentUserContext";
+import * as auth from "../utils/auth";
+import { addArticle, removeArticle } from "../utils/mainApi";
 
 function App() {
   const [activePopup, setActivePopup] = React.useState("");
   const [isLoggedIn, setIsLoggedIn] = React.useState(true);
-  //const [currentUser, setCurrentUser] = React.useState({});
-  //const [token, setToken] = React.useState("");
+  const [currentUser, setCurrentUser] = React.useState({});
+  const [token, setToken] = React.useState("");
   const [isLoading, setIsLoading] = React.useState(false);
   const [activeSearch, setActiveSearch] = React.useState(false);
   const [newsCards, setNewsCards] = React.useState([]);
   const [isSearchLoading, setIsSearchLoading] = React.useState(true);
+  const [errorMessage, setErrorMessage] = React.useState("");
+  const [keyword, setKeyword] = React.useState("");
+
+  const navigate = useNavigate();
+
+  const handleLogin = (email, pw) => {
+    auth
+      .signIn(email, pw)
+      .then((data) => {
+        if (data.token) {
+          auth
+            .checkToken(data.token)
+            .then((res) => {
+              return res;
+            })
+            .then((data) => {
+              setCurrentUser(data);
+            })
+            .then(() => {
+              setIsLoggedIn(true);
+            })
+            .then(() => {
+              navigate("/saved-articles");
+            })
+            .catch((err) => console.log(err));
+        }
+      })
+      .then(() => {
+        closePopups();
+      })
+      .catch((err) => {
+        console.log(err);
+        setErrorMessage("Username or password is incorrect");
+        setIsLoading(false);
```

Comment:
**Might be better**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js
  function handleSubmit(request) {
    // start loading
    setIsLoading(true);
    request()
      // we need to close only in `then`
      .then(closeAllPopups)
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
      .catch(console.error)
      // and in finally we need to stop loading
      .finally(() => setIsLoading(false));
  }

```


You need to pass the **request** into the call of `handleSubmit` like that:

```js
 // here is an example
 function handleProfileFormSubmit(inputValues) {
  // here we create a function that returns a promise 
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return api.editProfile(inputValues).then(setCurrentUser);
  }
  // here we call handleSubmit passing the request
  handleSubmit(makeRequest);
}


```
So, this way you can remove a lot of code duplicating